### ISSUE TITLE
Add check for unused module-level private and protected functions

### ISSUE THREAD

**Issue by socketpair** [PyCQA/pylint#4542]
test.py:
```python
def _unused():
    pass
def __unused():
    pass
```

This file should generate errors about unused private?protected functions. Just like private methods #4504

**Comment by DGarciaMedina** [comment]
I agree that this check would be very useful to have! Would allow developers to remove dead code that isn't referenced anywhere.

One thing to note is that the line with ```_some_constant = 42``` currently triggers the error ```Constant name "_some_constant" doesn't conform to UPPER_CASE naming style (invalid-name)```. In this case I believe it would be better to replace the existing message with something like what @socketpair was suggesting, since the variable name starts with ```_```, which indicates the intent of using this as a private variable.

I would be happy to create a PR for this.

**Comment by Pierre-Sassoulas** [comment]
Regarding the issue with the UPPER_CASE naming style it's another issue entirely and should be discussed elsewhere (I think the consensus is to allow both snake_case and upper case naming style for module level variable.

Regarding the unused level module private and protected function, please check what is currently being done in [this related MR](https://github.com/PyCQA/pylint/pull/5610) if you want to help :)

**Comment by socketpair** [comment]
@DGarciaMedina how do you do ?

**Comment by ollie-iterators** [comment]
You could use the python tool dead. It is in the repository: https://github.com/DeadCodeProductions/dead

**Comment by socketpair** [comment]
```
pylint 3.0.4
astroid 3.0.3
Python 3.12.2 (main, Feb 21 2024, 00:00:00) [GCC 13.2.1 20231205 (Red Hat 13.2.1-6)]
```

Still actual
